central/config/datastore/datastore_test.go

```diff
@@ -203,3 +208,274 @@ func (s *configDataStoreTestSuite) TestAllowsUpdate() {
+		},
+	}
+
+	customDecommissionedClusterRetention = &storage.DecommissionedClusterRetentionConfig{
+		RetentionDurationDays: DefaultDecommissionedClusterRetentionDays + 1,
+	}
+
+	customReportRetentionConfig = &storage.ReportRetentionConfig{
+		HistoryRetentionDurationDays:           DefaultReportHistoryRetentionWindow + 1,
+		DownloadableReportRetentionDays:        DefaultDownloadableReportRetentionDays + 1,
+		DownloadableReportGlobalRetentionBytes: DefaultDownloadableReportGlobalRetentionBytes + 1,
+	}
+
+	customAdministrationEventsConfig = &storage.AdministrationEventsConfig{
+		RetentionDurationDays: DefaultAdministrationEventsRetention + 1,
+	}
+)
+
+func TestValidateConfigAndPopulateMissingDefaults(t *testing.T) {
+	testCases := map[string]struct {
+		enabledFlags   []string
+		disabledFlags  []string
+		initialConfig  *storage.Config
+		upsertedConfig *storage.Config
+	}{
+		"No Update for fully set config": {
+			initialConfig: &storage.Config{
+				PublicConfig:  samplePublicConfig,
+				PrivateConfig: customPrivateConfig,
+			},
+			upsertedConfig: nil,
+		},
+		"Missing private config gets fully configured when Features activated": {
+			enabledFlags: []string{features.UnifiedCVEDeferral.EnvVar()},
+			initialConfig: &storage.Config{
+				PublicConfig:  samplePublicConfig,
+				PrivateConfig: nil,
+			},
+			upsertedConfig: &storage.Config{
+				PublicConfig: samplePublicConfig,
+				PrivateConfig: &storage.PrivateConfig{
+					AlertRetention:                      defaultAlertRetention,
+					ImageRetentionDurationDays:          DefaultImageRetention,
+					ExpiredVulnReqRetentionDurationDays: DefaultExpiredVulnReqRetention,
+					DecommissionedClusterRetention:      defaultDecommissionedClusterRetention,
+					ReportRetentionConfig:               defaultReportRetentionConfig,
+					VulnerabilityExceptionConfig:        defaultVulnerabilityDeferralConfig,
+					AdministrationEventsConfig:          defaultAdministrationEventsConfig,
+				},
+			},
+		},
+		"Missing private config gets partially configured when Features deactivated": {
+			disabledFlags: []string{features.UnifiedCVEDeferral.EnvVar()},
+			initialConfig: &storage.Config{
+				PublicConfig:  samplePublicConfig,
+				PrivateConfig: nil,
+			},
+			upsertedConfig: &storage.Config{
+				PublicConfig: samplePublicConfig,
+				PrivateConfig: &storage.PrivateConfig{
+					AlertRetention:                      defaultAlertRetention,
+					ImageRetentionDurationDays:          DefaultImageRetention,
+					ExpiredVulnReqRetentionDurationDays: DefaultExpiredVulnReqRetention,
+					DecommissionedClusterRetention:      defaultDecommissionedClusterRetention,
+					ReportRetentionConfig:               defaultReportRetentionConfig,
+					VulnerabilityExceptionConfig:        nil,
+					AdministrationEventsConfig:          defaultAdministrationEventsConfig,
+				},
+			},
+		},
+		"Configure decommissioned cluster retention when missing": {
+			initialConfig: &storage.Config{
```

Comment:
it's hard to read and diff initial and upserted config, but we do not have a builder that will solve this problem as we could change only fields that differs.